### PR TITLE
ENH: linalg: allow lstsq to work with 0-shaped arrays

### PR DESCRIPTION
**Pull request by akhmerov**
This is mostly to avoid dealing with corner cases in downstream code. The workaround is quite straightforward, but I thought it could be useful in scipy as well.

### CODE REVIEW

**Comment by rgommers** [comment]
``b`` is now explicitly documented as ``1-D or 2-D array``. Now you're accepting array scalars as well, so docs would need a tweak.

What about shapes ``(n, 0)`` rather than ``(0, n)``?

File: scipy/linalg/basic.py [hunk 1]
```diff
@@ -1084,6 +1084,8 @@ def lstsq(a, b, cond=None, overwrite_a=False, overwrite_b=False,
         nrhs = 1
     if m != b1.shape[0]:
         raise ValueError('incompatible dimensions')
+    if not m:  # Zero-sized problem, confuses LAPACK
```
Comment (nmayorov) [inline scipy/linalg/basic.py]: I think it is not recommended practice for integers, just `if m == 0:`.

Comment (akhmerov) [inline scipy/linalg/basic.py]: Fixed, thanks.

**Comment by akhmerov** [comment]
@rgommers I did not touch the pars that check the array dimensionality, so I don't see how my changes would make the code work with scalar. The docs do not mention that `N, M, K` have to be nonzero, and  mentioning that zero-sized arrays *also* work feels somewhat superfluous.

Thanks for the remark about `n=0`, I've updated the code to take this into account.

**Comment by WarrenWeckesser** [comment]
Needs tests for the case `n=0`, otherwise looks good to me.

**Comment by akhmerov** [comment]
Done.

**Comment by WarrenWeckesser** [comment]
I just noticed a problem.  `lstsq` must return *four* values: `x`, `residues`, `rank` and `s`.  Simply returning `np.zeros((n,) + b1.shape[1:], dtype=np.common_type(a1, b1))` is not correct.  (I haven't thought about what those values should be in the degenerate cases.)

**Comment by akhmerov** [comment]
@WarrenWeckesser ouch, indeed! I'll fix that. It's not so hard: `rank=0`, `residues=np.linalg.norm(b)`, `s=np.empty((0,))`

**Comment by akhmerov** [comment]
@WarrenWeckesser done.

I have also observed that the docstring was incorrect about the shape of residues. The empty array was not shape `()` (that's numpy scalar), but rather shape `(0,)`. If `b` had shape `(K,)`, the return value was a numpy scalar.

I have followed this convention, and I fixed the docstring in a separate commit.

**Comment by akhmerov** [comment]
Finally found the typo that made tests fail.

**Pull request merged**